osu.Game/Graphics/OsuColour.cs

```diff
@@ -229,6 +229,173 @@ public static Color4 ForegroundTextColourFor(Color4 backgroundColour)
             return Gray(brightness > 0.5f ? 0.2f : 0.9f);
         }
 
+        /// <summary>
+        /// Converts RGBA to HSPA.
+        /// </summary>
+        public static Color4 ToHSPA(Color4 colour)
+        {
+            const float p_r = 0.299f;
+            const float p_g = 0.587f;
+            const float p_b = 0.114f;
+
+            Color4 result = new Color4
+            {
+                A = colour.A,
+                B = MathF.Sqrt(colour.R * colour.R * p_r + colour.G * colour.G * p_g + colour.B + colour.B * p_b)
+            };
+
+            if (colour.R == colour.G && colour.R == colour.B)
+                return result;
+
+            if (colour.R >= colour.G && colour.R >= colour.B)
+            {
+                if (colour.B >= colour.G)
+                {
+                    result.R = 6f / 6f - 1f / 6f * (colour.B - colour.G) / (colour.R - colour.G);
+                    result.G = 1f - colour.G / colour.R;
+                }
+                else
+                {
+                    result.R = 0f / 6f + 1f / 6f * (colour.G - colour.B) / (colour.R - colour.B);
+                    result.G = 1f - colour.B / colour.R;
+                }
+            }
+            else if (colour.G >= colour.R && colour.G >= colour.B)
+            {
+                if (colour.R >= colour.B)
+                {
+                    result.R = 2f / 6f - 1f / 6f * (colour.R - colour.B) / (colour.G - colour.B);
+                    result.G = 1f - colour.B / colour.G;
+                }
+                else
+                {
+                    result.R = 2f / 6f + 1f / 6f * (colour.B - colour.R) / (colour.G - colour.R);
+                    result.G = 1f - colour.R / colour.G;
+                }
+            }
+            else
+            {
+                if (colour.G >= colour.R)
+                {
+                    result.R = 4f / 6f - 1f / 6f * (colour.G - colour.R) / (colour.B - colour.R);
+                    result.G = 1f - colour.R / colour.B;
+                }
+                else
+                {
+                    result.R = 4f / 6f + 1f / 6f * (colour.R - colour.G) / (colour.B - colour.G);
+                    result.G = 1f - colour.G / colour.B;
+                }
+            }
+
+            return result;
+        }
+
+        public static Color4 FromHSPA(Color4 colour)
+        {
+            const float p_r = 0.299f;
+            const float p_g = 0.587f;
+            const float p_b = 0.114f;
+
+            float minOverMax = 1f - colour.G;
+
+            Color4 result = new Color4 { A = colour.A };
+
+            if (minOverMax > 0f)
+            {
+                float part = 1f + colour.R * (1f / minOverMax - 1f);
```

Comment:
On comparing this with the reference implementation linked in OP, I don't think it's correct to extract this outside here. The original code has:

```
  if (minOverMax>0.) {
    if      ( H<1./6.) {   //  R>G>B
      H= 6.*( H-0./6.); part=1.+H*(1./minOverMax-1.);
    // etc
```

Note that `H` is assigned to *prior* to calculation of `part`, and the formula used to compute the new `H` is different in each of the 6 branches.

So I don't think this is right, and each branch has to compute `part` on its own after modifying `colour.R`. That said I can't say I see it making a huge amount of difference upon checking, though I haven't checked much more than one map:

| before | after |
| :-: | :-: |
| ![osu_2022-10-13_22-25-53](https://user-images.githubusercontent.com/20418176/195703886-8f79b448-0610-44a9-a676-2227881578d3.jpg) | ![osu_2022-10-13_22-25-15](https://user-images.githubusercontent.com/20418176/195703907-cc761d46-5a4a-4567-a3ef-dc19462cf669.jpg) |

Reply:
This is what happens when you need to make an entire website to explain your code style :facepalm: 

I'll go over it again because this isn't the first error I've made.